crda/remotematcher.go

```diff
@@ -0,0 +1,254 @@
+	}
+}
+
+// Name implements driver.Matcher.
+func (*Matcher) Name() string { return "crda" }
+
+// Filter implements driver.Matcher.
+func (*Matcher) Filter(record *claircore.IndexRecord) bool {
+	return record.Package.NormalizedVersion.Kind == "pep440"
+}
+
+// Query implements driver.Matcher.
+func (*Matcher) Query() []driver.MatchConstraint {
+	panic("unreachable")
+}
+
+// Vulnerable implements driver.Matcher.
+func (*Matcher) Vulnerable(ctx context.Context, record *claircore.IndexRecord, vuln *claircore.Vulnerability) (bool, error) {
+	// RemoteMatcher can match Package and Vulnerability.
+	panic("unreachable")
+}
+
+// QueryRemoteMatcher implements driver.RemoteMatcher.
+func (m *Matcher) QueryRemoteMatcher(ctx context.Context, records []*claircore.IndexRecord) (map[string][]*claircore.Vulnerability, error) {
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "crda/remotematcher/QueryRemoteMatcher").
+		Logger()
+	ctx = log.WithContext(ctx)
+	log.Debug().
+		Int("records", len(records)).
+		Msg("packages")
+
+	ctrlC, errorC := m.fetchVulnerabilities(ctx, records)
+	results := make(map[string][]*claircore.Vulnerability)
+	for r := range ctrlC {
+		for _, vuln := range r {
+			results[vuln.Package.ID] = append(results[vuln.Package.ID], vuln)
+		}
+	}
+	select {
+	case err := <-errorC: // guaranteed to have an err or be closed
+		// Don't propagate error, log and move on.
+		if err != nil {
+			log.Error().Err(err).Msg("call to component analyses has failed")
+		}
+	default:
+	}
+	log.Debug().
+		Int("vulnerabilities", len(results)).
+		Msg("query")
+	return results, nil
+}
+
+func (m *Matcher) fetchVulnerabilities(ctx context.Context, records []*claircore.IndexRecord) (chan []*claircore.Vulnerability, chan error) {
+	inC := make(chan *claircore.IndexRecord, m.requestConcurrency)
+	ctrlC := make(chan []*claircore.Vulnerability, m.requestConcurrency)
+	errorC := make(chan error, 1)
+	go func() {
+		defer close(errorC)
+		defer close(ctrlC)
+		var g errgroup.Group
+		for _, record := range records {
+			g.Go(func() error {
+				vulns, err := m.componentAnalyses(ctx, <-inC)
+				if err != nil {
+					return err
+				}
+				ctrlC <- vulns
+				return nil
+			})
+			inC <- record
+		}
+		close(inC)
+		if err := g.Wait(); err != nil {
+			errorC <- err
+		}
+	}()
+	return ctrlC, errorC
+}
+
+func (m *Matcher) componentAnalyses(ctx context.Context, record *claircore.IndexRecord) ([]*claircore.Vulnerability, error) {
+	reqUrl := url.URL{
+		Scheme:   m.url.Scheme,
+		Host:     m.url.Host,
+		Path:     fmt.Sprintf(defaultPath, record.Package.Name, record.Package.Version),
+		RawQuery: m.url.RawQuery,
+	}
+
+	req := http.Request{
+		Method:     http.MethodGet,
+		Header:     http.Header{"User-Agent": {"claircore/crda/RemoteMatcher"}},
+		URL:        &reqUrl,
+		Proto:      "HTTP/1.1",
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		Host:       reqUrl.Host,
+	}
+	// A request shouldn't go beyound 10s.
+	tctx, cancel := context.WithTimeout(ctx, 10*time.Second)
```

Comment:
@arajkumar would it be safe to roll this back to 1 second timeout? do you know how long an average API request to your api will take?

Reply:
@ldelossa On avg it should take between 2 - 3 seconds, How about 5 seconds?

Reply:
I can make it configurable.

Reply:
Lets keep it out of config for now, but 5sec sounds good

Reply:
Changed it to 5 sec.